src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts

```diff
@@ -0,0 +1,254 @@
+import { SimpleWorkerClient } from 'vs/base/common/worker/simpleWorker';
+
+export class LanguageDetectionService extends Disposable implements ILanguageDetectionService {
+	static readonly enablementSettingKey = 'workbench.editor.untitled.experimentalLanguageDetection';
+
+	_serviceBrand: undefined;
+
+	private _languageDetectionWorkerClient: LanguageDetectionWorkerClient;
+
+	constructor(
+		@IWorkbenchEnvironmentService private readonly _environmentService: IWorkbenchEnvironmentService,
+		@IModeService private readonly _modeService: IModeService,
+		@IConfigurationService private readonly _configurationService: IConfigurationService,
+		@IModelService modelService: IModelService,
+	) {
+		super();
+
+		this._languageDetectionWorkerClient = new LanguageDetectionWorkerClient(
+			modelService,
+			'languageDetectionWorkerService',
+			// TODO: See if it's possible to bundle vscode-languagedetection
+			FileAccess.asBrowserUri('../../../../../../node_modules/@vscode/vscode-languagedetection/dist/lib/index.js', require).toString(true),
+			this._environmentService.isBuilt && !isWeb
+				? FileAccess.asBrowserUri('../../../../../../node_modules.asar.unpacked/@vscode/vscode-languagedetection/model/model.json', require).toString(true)
+				: FileAccess.asBrowserUri('../../../../../../node_modules/@vscode/vscode-languagedetection/model/model.json', require).toString(true),
+			this._environmentService.isBuilt && !isWeb
+				? FileAccess.asBrowserUri('../../../../../../node_modules.asar.unpacked/@vscode/vscode-languagedetection/model/group1-shard1of1.bin', require).toString(true)
+				: FileAccess.asBrowserUri('../../../../../../node_modules/@vscode/vscode-languagedetection/model/group1-shard1of1.bin', require).toString(true));
+	}
+
+	public isEnabledForMode(modeId: string): boolean {
+		return !!modeId && this._configurationService.getValue<boolean>(LanguageDetectionService.enablementSettingKey, { overrideIdentifier: modeId });
+	}
+
+	private getModeId(language: string | undefined): string | undefined {
+		if (!language) {
+			return undefined;
+		}
+		return this._modeService.getModeIdByFilepathOrFirstLine(URI.file(`file.${language}`)) ?? undefined;
+	}
+
+	async detectLanguage(resource: URI): Promise<string | undefined> {
+		const language = await this._languageDetectionWorkerClient.detectLanguage(resource);
+		if (language) {
+			return this.getModeId(language);
+		}
+		return undefined;
+	}
+
+	async detectLanguages(resource: URI): Promise<string[]> {
+		const languages: Array<string | undefined> = await this._languageDetectionWorkerClient.detectLanguages(resource);
+		for (let i = 0; i < languages.length; i++) {
+			const modeId = this.getModeId(languages[i]);
+			languages[i] = modeId ? modeId : undefined;
+		}
+
+		return languages.filter(<T>(l?: T): l is T => Boolean(l));
+	}
+}
+
+export interface IWorkerClient<W> {
+	getProxyObject(): Promise<W>;
+	dispose(): void;
+}
+
+class LanguageDetectionModelManager extends Disposable {
```

Comment:
This was lifted from here:
https://github.com/microsoft/vscode/blob/025c9ce85fd369e668154c817d58098428cfe4b2/src/vs/editor/common/services/editorWorkerServiceImpl.ts#L276-L276

can I reuse it somehow?